### ISSUE TITLE
Voiceover

### ISSUE THREAD

**Issue by Dmitry-Demis** [philip-murray/reactive-manim#6]
Hello, I cannot use your library with manim.voiceover
        with self.voiceover(text="Have a look") as tracker:
            self.play(TransformInStages.progress(tex), run_time=tracker.duration)

```

from manim import *
from manim_voiceover import VoiceoverScene
from reactive_manim import *
from edge_tts_service import EdgeTTSService


class CasesScene(VoiceoverScene):
    def setup(self):
        # Настраиваем voiceover сервис
        self.set_speech_service(
            EdgeTTSService(
                voice="ru-RU-DmitryNeural",
                rate="+0%",
                pitch="+0Hz"
            )
        )
    
    def construct(self):
        tex = MathTex(
            "f(x) =",
            MathCases(
                CaseLine(r"b,     ", r"\text{even}(x)"),
                CaseLine(r"\neg b,", r"\text{odd}(x)"),
                CaseLine(r"?,", r"\text{otherwise}"),
            ),
        )
        
        self.add(tex).wait(1)
        
        cases = tex[1]
        cases[1].condition.set_color(ORANGE)
        cases[2].output.set_color(ORANGE)
        
        with self.voiceover(text="LOOK") as tracker:
            self.play(TransformInStages.progress(tex), run_time=tracker.duration)
        
        cases[1].condition = cases[2].condition
        cases.lines = [cases[0], cases[1]]
        
        
        self.wait(2)
```

```
 C:\Users\User\AppData\Roaming\uv\tools\manim\Lib\site-packages\manim\cli\render\commands.py:125  │
│ in render                                                                                        │
│                                                                                                  │
│   122 │   │   │   try:                                                                           │
│   123 │   │   │   │   with tempconfig({}):                                                       │
│   124 │   │   │   │   │   scene = SceneClass()                                                   │
│ ❱ 125 │   │   │   │   │   scene.render()                                                         │
│   126 │   │   │   except Exception:                                                              │
│   127 │   │   │   │   error_console.print_exception()                                            │
│   128 │   │   │   │   sys.exit(1)                                                                │
│                                                                                                  │
│ C:\Users\User\AppData\Roaming\uv\tools\manim\Lib\site-packages\manim\scene\scene.py:259 in       │
│ render                                                                                           │
│                                                                                                  │
│    256 │   │   """                                                                               │
│    257 │   │   self.setup()                                                                      │
│    258 │   │   try:                                                                              │
│ ❱  259 │   │   │   self.construct()                                                              │
│    260 │   │   except EndSceneEarlyException:                                                    │
│    261 │   │   │   pass                                                                          │
│    262 │   │   except RerunSceneException:                                                       │
│                                                                                                  │
│ D:\Studies\MANIM\test_module\test_scene.py:38 in construct                                       │
│                                                                                                  │
│   35 │   │   cases[1].condition.set_color(ORANGE)                                                │
│   36 │   │   cases[2].output.set_color(ORANGE)                                                   │
│   37 │   │                                                                                       │
│ ❱ 38 │   │   with self.voiceover(text="LOOK") as tracker:                                        │
│   39 │   │   │   self.play(TransformInStages.progress(tex), run_time=tracker.duration)           │
│   40 │   │                                                                                       │
│   41 │   │   cases[1].condition = cases[2].condition                                             │
│                                                                                                  │
│ C:\Users\User\AppData\Roaming\uv\python\cpython-3.13.11-windows-x86_64-none\Lib\contextlib.py:14 │
│ 1 in __enter__                                                                                   │
│                                                                                                  │
│   138 │   │   # they are only needed for recreation, which is not possible anymore               │
│   139 │   │   del self.args, self.kwds, self.func                                                │
│   140 │   │   try:                                                                               │
│ ❱ 141 │   │   │   return next(self.gen)                                                          │
│   142 │   │   except StopIteration:                                                              │
│   143 │   │   │   raise RuntimeError("generator didn't yield") from None                         │
│   144                                                                                            │
│                                                                                                  │
│ C:\Users\User\AppData\Local\Programs\Python\Python314\Lib\site-packages\manim_voiceover\voiceove │
│ r_scene.py:187 in voiceover                                                                      │
│                                                                                                  │
│   184 │   │                                                                                      │
│   185 │   │   try:                                                                               │
│   186 │   │   │   if text is not None:                                                           │
│ ❱ 187 │   │   │   │   yield self.add_voiceover_text(text, **kwargs)                              │
│   188 │   │   │   elif ssml is not None:                                                         │
│   189 │   │   │   │   yield self.add_voiceover_ssml(ssml, **kwargs)                              │
│   190 │   │   finally:                                                                           │
│                                                                                                  │
│ C:\Users\User\AppData\Local\Programs\Python\Python314\Lib\site-packages\manim_voiceover\voiceove │
│ r_scene.py:69 in add_voiceover_text                                                              │
│                                                                                                  │
│    66 │   │   │   │   "You need to call init_voiceover() before adding a voiceover."             │
│    67 │   │   │   )                                                                              │
│    68 │   │                                                                                      │
│ ❱  69 │   │   dict_ = self.speech_service._wrap_generate_from_text(text, **kwargs)               │
│    70 │   │   tracker = VoiceoverTracker(self, dict_, self.speech_service.cache_dir)             │
│    71 │   │   self.renderer.skip_animations = self.renderer._original_skipping_status            │
│    72 │   │   self.add_sound(str(Path(self.speech_service.cache_dir) / dict_["final_audio"]))    │
│                                                                                                  │
│ C:\Users\User\AppData\Local\Programs\Python\Python314\Lib\site-packages\manim_voiceover\services │
│ \base.py:90 in _wrap_generate_from_text                                                          │
│                                                                                                  │
│    87 │   │   # Replace newlines with lines, reduce multiple consecutive spaces to single        │
│    88 │   │   text = " ".join(text.split())                                                      │
│    89 │   │                                                                                      │
│ ❱  90 │   │   dict_ = self.generate_from_text(text, cache_dir=None, path=path, **kwargs)         │
│    91 │   │   original_audio = dict_["original_audio"]                                           │
│    92 │   │                                                                                      │
│    93 │   │   # Check whether word boundaries exist and if not run stt                           │
│                                                                                                  │
│ D:\Studies\MANIM\test_module\edge_tts_service.py:74 in generate_from_text                        │
│                                                                                                  │
│   71 │   │   # Запускаем в отдельном потоке чтобы избежать конфликта с существующим event loo    │
│   72 │   │   with ThreadPoolExecutor() as executor:                                              │
│   73 │   │   │   future = executor.submit(_run_in_thread)                                        │
│ ❱ 74 │   │   │   future.result()  # Ждём завершения                                              │
│   75 │   │                                                                                       │
│   76 │   │   # Проверяем что файл создан и имеет размер больше 0                                 │
│   77 │   │   max_wait = 5  # максимум 5 секунд ожидания                                          │
│                                                                                                  │
│ C:\Users\User\AppData\Roaming\uv\python\cpython-3.13.11-windows-x86_64-none\Lib\concurrent\futur │
│ es\_base.py:456 in result                                                                        │
│                                                                                                  │
│   453 │   │   │   │   if self._state in [CANCELLED, CANCELLED_AND_NOTIFIED]:                     │
│   454 │   │   │   │   │   raise CancelledError()                                                 │
│   455 │   │   │   │   elif self._state == FINISHED:                                              │
│ ❱ 456 │   │   │   │   │   return self.__get_result()                                             │
│   457 │   │   │   │   else:                                                                      │
│   458 │   │   │   │   │   raise TimeoutError()                                                   │
│   459 │   │   finally:                                                                           │
│                                                                                                  │
│ C:\Users\User\AppData\Roaming\uv\python\cpython-3.13.11-windows-x86_64-none\Lib\concurrent\futur │
│ es\_base.py:401 in __get_result                                                                  │
│                                                                                                  │
│   398 │   def __get_result(self):                                                                │
│   399 │   │   if self._exception is not None:                                                    │
│   400 │   │   │   try:                                                                           │
│ ❱ 401 │   │   │   │   raise self._exception                                                      │
│   402 │   │   │   finally:                                                                       │
│   403 │   │   │   │   # Break a reference cycle with the exception in self._exception            │
│   404 │   │   │   │   self = None                                                                │
│                                                                                                  │
│ C:\Users\User\AppData\Roaming\uv\python\cpython-3.13.11-windows-x86_64-none\Lib\concurrent\futur │
│ es\thread.py:59 in run                                                                           │
│                                                                                                  │
│    56 │   │   │   return                                                                         │
│    57 │   │                                                                                      │
│    58 │   │   try:                                                                               │
│ ❱  59 │   │   │   result = self.fn(*self.args, **self.kwargs)                                    │
│    60 │   │   except BaseException as exc:                                                       │
│    61 │   │   │   self.future.set_exception(exc)                                                 │
│    62 │   │   │   # Break a reference cycle with the exception 'exc'                             │
│                                                                                                  │
│ D:\Studies\MANIM\test_module\edge_tts_service.py:67 in _run_in_thread                            │
│                                                                                                  │
│   64 │   │   │   new_loop = asyncio.new_event_loop()                                             │
│   65 │   │   │   asyncio.set_event_loop(new_loop)                                                │
│   66 │   │   │   try:                                                                            │
│ ❱ 67 │   │   │   │   new_loop.run_until_complete(_async_generate())                              │
│   68 │   │   │   finally:                                                                        │
│   69 │   │   │   │   new_loop.close()                                                            │
│   70                                                                                             │
│                                                                                                  │
│ C:\Users\User\AppData\Roaming\uv\python\cpython-3.13.11-windows-x86_64-none\Lib\asyncio\base_eve │
│ nts.py:725 in run_until_complete                                                                 │
│                                                                                                  │
│    722 │   │   if not future.done():                                                             │
│    723 │   │   │   raise RuntimeError('Event loop stopped before Future completed.')             │
│    724 │   │                                                                                     │
│ ❱  725 │   │   return future.result()                                                            │
│    726 │                                                                                         │
│    727 │   def stop(self):                                                                       │
│    728 │   │   """Stop running the event loop.                                                   │
│                                                                                                  │
│ D:\Studies\MANIM\test_module\edge_tts_service.py:61 in _async_generate                           │
│                                                                                                  │
│   58 │   │   │   │   │   rate=self.rate,                                                         │
│   59 │   │   │   │   │   pitch=self.pitch                                                        │
│   60 │   │   │   │   )                                                                           │
│ ❱ 61 │   │   │   │   await communicate.save(audio_path)                                          │
│   62 │   │   │                                                                                   │
│   63 │   │   │   # Создаём новый event loop в отдельном потоке                                   │
│   64 │   │   │   new_loop = asyncio.new_event_loop()                                             │
│                                                                                                  │
│ C:\Users\User\AppData\Local\Programs\Python\Python314\Lib\site-packages\edge_tts\communicate.py: │
│ 590 in save                                                                                      │
│                                                                                                  │
│   587 │   │   │   else nullcontext()                                                             │
│   588 │   │   )                                                                                  │
│   589 │   │   with metadata, open(audio_fname, "wb") as audio:                                   │
│ ❱ 590 │   │   │   async for message in self.stream():                                            │
│   591 │   │   │   │   if message["type"] == "audio":                                             │
│   592 │   │   │   │   │   audio.write(message["data"])                                           │
│   593 │   │   │   │   elif isinstance(metadata, TextIOWrapper) and message["type"] in (          │
│                                                                                                  │
│ C:\Users\User\AppData\Local\Programs\Python\Python314\Lib\site-packages\edge_tts\communicate.py: │
│ 566 in stream                                                                                    │
│                                                                                                  │
│   563 │   │   # Stream the audio and metadata from the service.                                  │
│   564 │   │   for self.state["partial_text"] in self.texts:                                      │
│   565 │   │   │   try:                                                                           │
│ ❱ 566 │   │   │   │   async for message in self.__stream():                                      │
│   567 │   │   │   │   │   yield message                                                          │
│   568 │   │   │   except aiohttp.ClientResponseError as e:                                       │
│   569 │   │   │   │   if e.status != 403:                                                        │
│                                                                                                  │
│ C:\Users\User\AppData\Local\Programs\Python\Python314\Lib\site-packages\edge_tts\communicate.py: │
│ 436 in __stream                                                                                  │
│                                                                                                  │
│   433 │   │   │   trust_env=True,                                                                │
│   434 │   │   │   timeout=self.session_timeout,                                                  │
│   435 │   │   ) as session, session.ws_connect(                                                  │
│ ❱ 436 │   │   │   f"{WSS_URL}&ConnectionId={connect_id()}"                                       │
│   437 │   │   │   f"&Sec-MS-GEC={DRM.generate_sec_ms_gec()}"                                     │
│   438 │   │   │   f"&Sec-MS-GEC-Version={SEC_MS_GEC_VERSION}",                                   │
│   439 │   │   │   compress=15,                                                                   │
│                                                                                                  │
│ C:\Users\User\AppData\Local\Programs\Python\Python314\Lib\site-packages\edge_tts\communicate.py: │
│ 100 in connect_id                                                                                │
│                                                                                                  │
│    97 │   Returns:                                                                               │
│    97 │   Returns:                                                                               │
│    98 │   │   str: A UUID without dashes.                                                        │
│    99 │   """                                                                                    │
│ ❱ 100 │   return uuid.uuid4().hex                                                                │
│   101                                                                                            │
│   102                                                                                            │
│   103 def _find_last_newline_or_space_within_limit(text: bytes, limit: int) -> int:              │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: 'int' object has no attribute 'hex'
```

**Comment by philip-murray** [comment]
I have uploaded a new version to PyPI, can you reinstall reactive-manim and post whether it works or not?

I am unable to run the code exactly as you have written it, because I'm not sure where you are getting `edge_tts_service`  from. However, I believe I have fixed the issue based off the stack-trace, so it should work with the updated version.
